### PR TITLE
refactor(suggestion): Use interface

### PR DESCRIPTION
**Pull request by gaocegege**
There are some benefits to define an interface instead of using util.GetSuggestions function:

- Easy to test and maintain. We can write unit test cases for the interface.
- Easy to integrate. We can mock the suggestion when testing the experiment controller, or use a fake controller to run the experiment without affecting the real logic.

This PR:

- added a CLI argument to control if we use fake implementation in the controller. If not, now we will always return an error becuase we do not implement suggestion logic now.

```
  -fake-experiment-suggestion
    	Use the fake suggestion in experiment controller
````

- added a interface about suggestion.

Signed-off-by: Ce Gao <gaoce@caicloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/502)

### CODE REVIEW

**Comment by gaocegege** [comment]
/cc @johnugeorge @richardsliu @YujiOshima 

PTAL

File: pkg/controller/v1alpha2/experiment/experiment_controller.go [hunk 1]
```diff
@@ -36,12 +37,15 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/builder"
 
-	experimentsv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/experiment/v1alpha2"
-	trialsv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/trial/v1alpha2"
+	experimentv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/experiment/v1alpha2"
```
Comment (johnugeorge) [inline pkg/controller/v1alpha2/experiment/experiment_controller.go]: I feel, Kubebuilder recommends plural form of CR.  Ideally, it should have been `apis/experiments`, `apis/trials` as per Kubebuilder notation. I was planning to take this up post 0.6 release

Comment (gaocegege) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: @johnugeorge SGTM. I reverted the change. Maybe we could rename the package in #495

Comment (johnugeorge) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: Sounds good

File: cmd/katib-controller/v1alpha2/main.go [hunk 2]
```diff
@@ -20,49 +20,70 @@ limitations under the License.
...
 )
 
 func main() {
 	logf.SetLogger(logf.ZapLogger(false))
+	log := logf.Log.WithName("entrypoint")
+
+	var fakeExperimentSuggestion bool
+
+	flag.BoolVar(&fakeExperimentSuggestion, "fake-experiment-suggestion",
+		false, "Use the fake suggestion in experiment controller")
```
Comment (hougangliu) [inline cmd/katib-controller/v1alpha2/main.go]: how about use args like `experiment-suggestion-name`, and now its value can be "fake|default", `default` by default. in future, we can may support more.

Comment (gaocegege) [reply cmd/katib-controller/v1alpha2/main.go]: Good point. I will update it.

Comment (gaocegege) [reply cmd/katib-controller/v1alpha2/main.go]: Addressed

**Comment by gaocegege** [comment]
@hougangliu PTAL

File: pkg/controller/v1alpha2/experiment/experiment_controller.go [hunk 3]
```diff
@@ -58,7 +62,21 @@ func Add(mgr manager.Manager) error {
...
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+	// If the flag is set in CLI, use the fake implementation.
+	imp := viper.GetString(consts.ConfigFakeExperimentSuggestion)
+	switch imp {
+	case "fake":
+		log.Info("Using the fake suggestion implementation")
+		r.Suggestion = suggestionfake.New()
+	default:
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/experiment_controller.go]: if user sets it as "invalid-one", we can still use default, but a warning log is necessary

Comment (gaocegege) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: SGTM

**Comment by hougangliu** [comment]
> @hougangliu PTAL
@gaocegege Thanks. some mini comment

**Comment by gaocegege** [comment]
> > @hougangliu PTAL
> > @gaocegege Thanks. some mini comment

Addressed, Thanks for your review. PTAL

**Comment by hougangliu** [comment]
/lgtm
thanks

File: pkg/controller/v1alpha2/experiment/experiment_controller.go [hunk 4]
```diff
@@ -58,7 +62,30 @@ func Add(mgr manager.Manager) error {
...
+		scheme: mgr.GetScheme(),
+	}
+	imp := viper.GetString(consts.ConfigExperimentSuggestionName)
+	r.Suggestion = newSuggestion(imp)
+	return r
+}
+
+// newSuggestion returns the new Suggestion for the given config.
+func newSuggestion(implementation string) suggestion.Suggestion {
+	// If the flag is set in CLI, use the fake implementation.
```
Comment (YujiOshima) [inline pkg/controller/v1alpha2/experiment/experiment_controller.go]: The comment is outdated.

Comment (gaocegege) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: Thanks for your review, addressed, PTAL again https://github.com/kubeflow/katib/pull/502/commits/04b2032fc379203cca0a5f93a33ef7eda154d35d

**Comment by gaocegege** [comment]
@hougangliu PTAL again, fixed a bug.

File: pkg/controller/v1alpha2/experiment/experiment_controller.go [hunk 5]
```diff
@@ -58,7 +62,30 @@ func Add(mgr manager.Manager) error {
 
 // newReconciler returns a new reconcile.Reconciler
 func newReconciler(mgr manager.Manager) reconcile.Reconciler {
-	return &ReconcileExperiment{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+	r := &ReconcileExperiment{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+	imp := viper.GetString(consts.ConfigExperimentSuggestionName)
```
Comment (johnugeorge) [inline pkg/controller/v1alpha2/experiment/experiment_controller.go]: Just to check: Is there a better way to pass flags to the manager than using viper?

Comment (gaocegege) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: I have no idea. Kubebuilder does not support a built-in way, thus I use viper here.

Comment (gaocegege) [reply pkg/controller/v1alpha2/experiment/experiment_controller.go]: Does it work for you?

**Pull request merged**